File: 0301-0400/0396-Rotate Function/0396-Rotate Function.cpp
class Solution {
public:
    int maxRotateFunction(vector<int>& A) {
        int n = A.size();
        int sum = 0;
        int curr = 0;
        for (int i = 0; i < n; ++i) {
            sum += A[i];
            curr += i * A[i];
        }

        int maxVal = curr;
        for (int k = 1; k < n; ++k) {
            curr += sum - n * A[n - k];
            maxVal = max(maxVal, curr);
        }
        
        return maxVal;
    }
};

File: 0301-0400/0396-Rotate Function/0396-Rotate Function.py
class Solution:
    def maxRotateFunction(self, A: 'List[int]') -> 'int':
        s = sum(A)
        curr = sum(i*a for i, a in enumerate(A))
        maxVal = curr
        for i in range(1, len(A)):
            curr += s - len(A)*A[-i]
            maxVal = max(maxVal, curr)
        return maxVal
